### DISCUSSION TITLE
Amend workflow problems

### DISCUSSION THREAD

**Discussion by alan-nf** [arxanas/git-branchless#324] [Q&A] [answered]
I'm having some trouble with `amend` and restacking in my real repository (note: it works totally fine in my test repo).

Here are my issues:
* If I `git commit --amend`, branchless doesn't know the commit is rewritten (it doesn't say it's rewritten in `git sl`)
* If I `git amend` I get an error that the command doesn't exist. I had to manually alias it. (I saw a PR where you stopped adding this alias, so this probably just a needed [documentation](https://github.com/arxanas/git-branchless/wiki/Command:-git-amend#usage) update. If so let me know and I can edit the wiki page myself)
* If I `git branchless amend` (or add an `amend` alias) then it works, and `git sl` reports the rewritten commit, but a `restack` is not attempted.
* If I manually `git restack` it says there are no abandoned commits to restack. Aren't descendants of rewritten commits abandoned, or am I missing something? Here's the output:
```
➜  myrepo git:(b725e8d) ✗ git branchless amend
branchless: running command: git reset
branchless: processing 1 update: ref HEAD
No abandoned commits to restack.
No abandoned branches to restack.
⋮
◇ e994b6ca 1d (main, remote origin/main) <redacted>
┣━┓
┃ ◯ b4fced5b 23h (alan/feature1) <redacted>
┃ ┃
┃ ◯ d5272462 23h <redacted>
┃ ┃
┃ ◯ c56a74a2 23h <redacted>
┃ ┃
┃ ◯ fa4c6ddb 23h <redacted>
┃ ┣━┓
┃ ┃ ✕ b725e8d6 23h (rewritten as 06c16bb8) <redacted>     # rewritten commit
┃ ┃ ┃
┃ ┃ ◯ e1a9a7b5 23h <redacted>    # abandoned commits start here
┃ ┃ ┃
┃ ┃ ◯ 1576504d 23h <redacted>
┃ ┃ ┃
┃ ┃ ◯ 9a29234f 23h <redacted>
┃ ┃ ┃
┃ ┃ ◯ fe64f19b 23h <redacted>
┃ ┃ ┃
┃ ┃ ◯ c4c76e98 23h (alan/feature2) <redacted>
┃ ┃ ┃
┃ ┃ ◯ bd9a47f5 21h <redacted>
┃ ┃ ┃
┃ ┃ ◯ 04715177 21h <redacted>
┃ ┃ ┃
┃ ┃ ◯ f8200915 16h (alan/feature3) <redacted>
┃ ┃
┃ ● 06c16bb8 1s <redacted>   # new commit with the amended changes
┃
◯ b28977e1 23h (alan/some-other-feature) <redacted>
Amended with 1 uncommitted change.
➜  myrepo git:(06c16bb) git branchless restack
No abandoned commits to restack.
No abandoned branches to restack.
⋮
◇ e994b6ca 1d (main, remote origin/main) <redacted>
...
```

If I'm not misunderstanding this, is there a way I can troubleshoot what's going on? Like I said, this works in my test repo and is just not working in my main repo. Also note, I tried deleting the sqlite db and re-`init`ing and that didn't help: the behavior remains the same.

Lastly, I was able to work around this by just doing a `git move -s e1a9a7b5 --merge && git hide -r b725e8d6` which fixed me right up.

**Comment by arxanas** [comment, marked as answer]
Can you run `git branchless bug-report` (on v0.3.10+) after doing your failing amend and paste the output here?

**Reply by alan-nf** [reply]
Thanks for your responsive help, @arxanas. Much appreciated.

# I did:
```
cargo install --locked git-branchless
git branchless -V # 0.3.10
git branchless init --uninstall
git branchless init
git co 06c16bb
echo "ignore me, don't commit" >> test.txt
git add test.txt
git status
git branchless amend
git branchless bug-report
```

# And got:
#### Software version

git-branchless 0.3.10

#### Operating system

macOS 12.2.1 (Darwin 21.3.0)

#### Command-line

```bash
/Users/anorbauer/.cargo/bin/git-branchless bug-report 
```

#### Environment variables

```bash
SHELL=/bin/zsh
EDITOR=<not set>
```

#### Git version

```
> git version 
git version 2.32.0 (Apple Git-132)
```

#### Events


<details>
<summary>Show 5 events</summary>

##### Event ID: 71, transaction ID: 174

1. `RewriteEvent { timestamp: 1648581023.983188, event_tx_id: EventTransactionId(174), old_commit_oid: 06c16bb84cb99399cd20a6af83763e2300888e2c, new_commit_oid: 3ed00125affd48e8ecd84b7da6447bd1dcbde9bb }`
```
:
O e994b6c 1d (redacted-ref-0, remote origin/main) xxx xxxxxxxx x xxxxxx xx xxxxxxxxxxx xx xxxx xxxxxxxxxx
|\
| o b4fced5 23h (redacted-ref-1) xxxx xxxxx xxxx xxxxxxxxxxxx x xxxx xxxxx
| |
| o d527246 23h xxxx xxxxx xxx xx xxx xxxxxxxxxxxx
| |
| o c56a74a 23h xxxxxxx xxxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| |
| o fa4c6dd 23h xxxxxxx xxxxxxxxxxxxxxxxx xx xxx xxxxx
| |\
| | x 06c16bb 27m (rewritten as 3ed00125) xxxx xxxxx xxxxxx xxxxxx x xxxxx
| | |
| | o 27a1de2 23m xxxx xxxxx xxxx xxxxxxx xx xxxx xxxxxx xxxx
| | |
| | o 860f648 23m xxxxxxx xxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| | |
| | o b3c6c10 23m xxxxxx xxxxxxxxxxxxxxxxx xx xxxxxxxx xxx x xxxxxxxx
| | |
| | o e6baabc 23m xxx xxxxxx xxxx xxxxxxxxx
| | |
| | o dfc8498 23m (redacted-ref-2) xxxx xxxxxxxxxxxxxxxxxxx xxxx xxxx x xxxxxx
| | |
| | o 3431ef2 23m xxxxxxx xxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxxx
| | |
| | o 0ae8166 23m xxx x xxxxx xxxxxxx
| | |
| | o 685e36f 23m (redacted-ref-3) xxxxxxx xxx xx xxxxxxxxxxxxxx
| |
| @ 3ed0012 45s xxxx xxxxx xxxxxx xxxxxx x xxxxx
|
o b28977e 23h (redacted-ref-4) xxx x xxxxx xxxxxx xxxxxxxxxxxxxxx xxxxxx
```
##### Event ID: 70, transaction ID: 171

1. `RefUpdateEvent { timestamp: 1648580996.763745, event_tx_id: EventTransactionId(171), ref_name: "HEAD", old_oid: 0000000000000000000000000000000000000000, new_oid: 06c16bb84cb99399cd20a6af83763e2300888e2c, message: None }`
```
:
O e994b6c 1d (redacted-ref-0, remote origin/main) xxx xxxxxxxx x xxxxxx xx xxxxxxxxxxx xx xxxx xxxxxxxxxx
|\
| o b4fced5 23h (redacted-ref-1) xxxx xxxxx xxxx xxxxxxxxxxxx x xxxx xxxxx
| |
| o d527246 23h xxxx xxxxx xxx xx xxx xxxxxxxxxxxx
| |
| o c56a74a 23h xxxxxxx xxxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| |
| o fa4c6dd 23h xxxxxxx xxxxxxxxxxxxxxxxx xx xxx xxxxx
| |\
| | x 06c16bb 27m xxxx xxxxx xxxxxx xxxxxx x xxxxx
| | |
| | o 27a1de2 23m xxxx xxxxx xxxx xxxxxxx xx xxxx xxxxxx xxxx
| | |
| | o 860f648 23m xxxxxxx xxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| | |
| | o b3c6c10 23m xxxxxx xxxxxxxxxxxxxxxxx xx xxxxxxxx xxx x xxxxxxxx
| | |
| | o e6baabc 23m xxx xxxxxx xxxx xxxxxxxxx
| | |
| | o dfc8498 23m (redacted-ref-2) xxxx xxxxxxxxxxxxxxxxxxx xxxx xxxx x xxxxxx
| | |
| | o 3431ef2 23m xxxxxxx xxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxxx
| | |
| | o 0ae8166 23m xxx x xxxxx xxxxxxx
| | |
| | o 685e36f 23m (redacted-ref-3) xxxxxxx xxx xx xxxxxxxxxxxxxx
| |
| @ 3ed0012 45s xxxx xxxxx xxxxxx xxxxxx x xxxxx
|
o b28977e 23h (redacted-ref-4) xxx x xxxxx xxxxxx xxxxxxxxxxxxxxx xxxxxx
```
##### Event ID: 55, transaction ID: 159

1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(e1a9a7b5762a7635b760690afe9fec7a5d843d6a) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(1576504d11043b1478d3820599f1cde0c310fa85) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(9a29234f3be67a660fde75b1192423836adc73f9) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(fe64f19b3fa00a5ee10f3c13d073cf83549b1d4d) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(c4c76e981d4dd1cb84c203fe23daceece6692c50) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(e542cb9d275eb797419bd5abbaaf294922e420d7) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(8ee97db7b4d7876ee8b49f0ae6564af2624e6ec0) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(17436dad799df644d0874f8b8e9098c912d18a13) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(bd9a47f587013036fcf64e3e49167c4fbe5e9665) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(2bcab3fd9b0ba214b9e1354c86a011ce7506bed5) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(04715177098ecc4c4c8e6055f1ebf3350043cfff) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(54de99c607aaa45d6099e011af2de0cc04fb4629) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(5b25dc521ae1d08b1180ef0ad2fe43a8d1ae57a1) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(3d12444607a12ab53a12a5a78deccbc308e91dc3) }`
1. `ObsoleteEvent { timestamp: 1648579676.127951, event_tx_id: EventTransactionId(159), commit_oid: NonZeroOid(f8200915697f764d5c4b427540ce3e805483d4b3) }`
```
:
O e994b6c 1d (redacted-ref-0, remote origin/main) xxx xxxxxxxx x xxxxxx xx xxxxxxxxxxx xx xxxx xxxxxxxxxx
|\
| o b4fced5 23h (redacted-ref-1) xxxx xxxxx xxxx xxxxxxxxxxxx x xxxx xxxxx
| |
| o d527246 23h xxxx xxxxx xxx xx xxx xxxxxxxxxxxx
| |
| o c56a74a 23h xxxxxxx xxxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| |
| o fa4c6dd 23h xxxxxxx xxxxxxxxxxxxxxxxx xx xxx xxxxx
| |\
| | x 06c16bb 27m xxxx xxxxx xxxxxx xxxxxx x xxxxx
| | |
| | o 27a1de2 23m xxxx xxxxx xxxx xxxxxxx xx xxxx xxxxxx xxxx
| | |
| | o 860f648 23m xxxxxxx xxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| | |
| | o b3c6c10 23m xxxxxx xxxxxxxxxxxxxxxxx xx xxxxxxxx xxx x xxxxxxxx
| | |
| | o e6baabc 23m xxx xxxxxx xxxx xxxxxxxxx
| | |
| | o dfc8498 23m (redacted-ref-2) xxxx xxxxxxxxxxxxxxxxxxx xxxx xxxx x xxxxxx
| | |
| | o 3431ef2 23m xxxxxxx xxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxxx
| | |
| | o 0ae8166 23m xxx x xxxxx xxxxxxx
| | |
| | o 685e36f 23m (redacted-ref-3) xxxxxxx xxx xx xxxxxxxxxxxxxx
| |
| @ 3ed0012 45s xxxx xxxxx xxxxxx xxxxxx x xxxxx
|
o b28977e 23h (redacted-ref-4) xxx x xxxxx xxxxxx xxxxxxxxxxxxxxx xxxxxx
```
##### Event ID: 54, transaction ID: 156

1. `ObsoleteEvent { timestamp: 1648579649.400911, event_tx_id: EventTransactionId(156), commit_oid: NonZeroOid(b725e8d6b0032d424bae6f7c6a7bb9109c47b691) }`
```
:
O e994b6c 1d (redacted-ref-0, remote origin/main) xxx xxxxxxxx x xxxxxx xx xxxxxxxxxxx xx xxxx xxxxxxxxxx
|\
| o b4fced5 23h (redacted-ref-1) xxxx xxxxx xxxx xxxxxxxxxxxx x xxxx xxxxx
| |
| o d527246 23h xxxx xxxxx xxx xx xxx xxxxxxxxxxxx
| |
| o c56a74a 23h xxxxxxx xxxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| |
| o fa4c6dd 23h xxxxxxx xxxxxxxxxxxxxxxxx xx xxx xxxxx
| |\
| | x 06c16bb 27m xxxx xxxxx xxxxxx xxxxxx x xxxxx
| | |
| | o 27a1de2 23m xxxx xxxxx xxxx xxxxxxx xx xxxx xxxxxx xxxx
| | |
| | o 860f648 23m xxxxxxx xxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| | |
| | o b3c6c10 23m xxxxxx xxxxxxxxxxxxxxxxx xx xxxxxxxx xxx x xxxxxxxx
| | |
| | o e6baabc 23m xxx xxxxxx xxxx xxxxxxxxx
| | |
| | o dfc8498 23m (redacted-ref-2) xxxx xxxxxxxxxxxxxxxxxxx xxxx xxxx x xxxxxx
| | |
| | o 3431ef2 23m xxxxxxx xxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxxx
| | |
| | o 0ae8166 23m xxx x xxxxx xxxxxxx
| | |
| | o 685e36f 23m (redacted-ref-3) xxxxxxx xxx xx xxxxxxxxxxxxxx
| |
| @ 3ed0012 45s xxxx xxxxx xxxxxx xxxxxx x xxxxx
|
o b28977e 23h (redacted-ref-4) xxx x xxxxx xxxxxx xxxxxxxxxxxxxxx xxxxxx
```
##### Event ID: 52, transaction ID: 154

1. `RefUpdateEvent { timestamp: 1648579640.74783, event_tx_id: EventTransactionId(154), ref_name: "refs/heads/redacted-ref-2", old_oid: c4c76e981d4dd1cb84c203fe23daceece6692c50, new_oid: dfc849865b657d0f350dbd255b0d8e63ec4d48f0, message: None }`
1. `RefUpdateEvent { timestamp: 1648579640.74783, event_tx_id: EventTransactionId(154), ref_name: "refs/heads/redacted-ref-3", old_oid: f8200915697f764d5c4b427540ce3e805483d4b3, new_oid: 685e36fee4ad83d306cad7cea043d684913fbbe7, message: None }`
```
:
O e994b6c 1d (redacted-ref-0, remote origin/main) xxx xxxxxxxx x xxxxxx xx xxxxxxxxxxx xx xxxx xxxxxxxxxx
|\
| o b4fced5 23h (redacted-ref-1) xxxx xxxxx xxxx xxxxxxxxxxxx x xxxx xxxxx
| |
| o d527246 23h xxxx xxxxx xxx xx xxx xxxxxxxxxxxx
| |
| o c56a74a 23h xxxxxxx xxxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| |
| o fa4c6dd 23h xxxxxxx xxxxxxxxxxxxxxxxx xx xxx xxxxx
| |\
| | x 06c16bb 27m xxxx xxxxx xxxxxx xxxxxx x xxxxx
| | |
| | o 27a1de2 23m xxxx xxxxx xxxx xxxxxxx xx xxxx xxxxxx xxxx
| | |
| | o 860f648 23m xxxxxxx xxxxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxx
| | |
| | o b3c6c10 23m xxxxxx xxxxxxxxxxxxxxxxx xx xxxxxxxx xxx x xxxxxxxx
| | |
| | o e6baabc 23m xxx xxxxxx xxxx xxxxxxxxx
| | |
| | o dfc8498 23m (redacted-ref-2) xxxx xxxxxxxxxxxxxxxxxxx xxxx xxxx x xxxxxx
| | |
| | o 3431ef2 23m xxxxxxx xxxxxxxxxxxxxxxxxx xx xxxxxxxxxxxxxxxxxxxxxx
| | |
| | o 0ae8166 23m xxx x xxxxx xxxxxxx
| | |
| | o 685e36f 23m (redacted-ref-3) xxxxxxx xxx xx xxxxxxxxxxxxxx
| |
| @ 3ed0012 45s xxxx xxxxx xxxxxx xxxxxx x xxxxx
|
o b28977e 23h (redacted-ref-4) xxx x xxxxx xxxxxx xxxxxxxxxxxxxxx xxxxxx
```

</details>

**Reply by arxanas** [reply]
I think this is a bug, and your understanding of concepts is correct.

> If I git amend I get an error that the command doesn't exist. I had to manually alias it

In the latest release, if you run `git branchless init` again, it should install the alias, but only if you don't already have one.

> Aren't descendants of rewritten commits abandoned, or am I missing something?

Nope, that's pretty much the definition.

My guess is that the fix for https://github.com/arxanas/git-branchless/issues/209 is inadequate in this case. Before, we were restacking too many commits, and now we're restacking too few. (Ironically, that's the issue which motivated the creation of a `bug-report` command to begin with.)

If the commits in question existed before git-branchless observed them (either via the `post-commit` or `post-rewrite` hook), then it could explain this behavior, as they would be missed in this line:

https://github.com/arxanas/git-branchless/blob/7073ea1937210b9b64ab34e5d82ada3a081d0f35/src/core/rewrite/evolve.rs#L85

I'll try to get around to fixing this sometime. (Filed #325.) In the meantime, `git move`/`git hide` should work as a workaround. You should notice that `git restack` starts working if the abandoned commits were created since the time you ran `git branchless init`.

**Reply by arxanas** [reply]
Fortunately, reproducing it in a test case turned out to be easy (added one line from the test case from #209). @alan-nf you can try a build including https://github.com/arxanas/git-branchless/pull/326 and see if it resolves your issue.

**Reply by alan-nf** [reply]
> You should notice that git restack starts working if the abandoned commits were created since the time you ran git branchless init.

Rebasing the commits in question (after `git branchless init`) doesn't seem to make it start working. Should that count as "created"? Is there some way for me to poke existing commits so they start getting tracked?

> Fortunately, reproducing it in a test case turned out to be easy (added one line from the test case from #209). @alan-nf you can try a build including #326 and see if it resolves your issue.

Thanks for the potential fix, @arxanas. I'd be happy to test it. How would I go about testing a pull request build? Do I pull the repo on that `restack` branch and then `cargo install` from it directly somehow? And do I need to re-`init` branchless after getting the new build?

**Reply by claytonrcarter** [reply]
You can either build and install the binary "globally", or you can just build the binary and then run it w/o installing it. (At least, that's what I've been doing as I've been hacking around.)

```
# clone; checkout; then ...

# either install "globally" via ...
$ cargo install --path .
$ cd my-repo
$ git restack ... 

# or just build it and run it locally via ...
$ cargo build
$ cd my-repo
$ /path/to/git-branchless-repo/target/debug/git-branchless restack ...
```

I have been doing the former when I have a test feature I want to play with in real work, and the latter when I just want to test something one-off.

**Reply by alan-nf** [reply]
@claytonrcarter perfect thanks. I [added that](https://github.com/arxanas/git-branchless/wiki/Runbook/_compare/177a50d2790a99a34039c7e7f2758a10de57443a) to a new section in the [wiki development runbook](https://github.com/arxanas/git-branchless/wiki/Runbook#manually-testing-from-the-git-repo)

**Reply by alan-nf** [reply]
@arxanas I just tested https://github.com/arxanas/git-branchless/pull/326 and it seemed to work perfectly. `git amend` successfully amended and restacked on its own. Thank you!

**Reply by arxanas** [reply]
@alan-nf thanks for updating the Wiki! For future reference, you can install from a remote branch directly like this:

```
$ cargo install --locked --git https://github.com/arxanas/git-branchless --branch restack
```

I updated the runbook with that information. @claytonrcarter I also added notes on using `cargo run`, which might simplify your workflow.
